File: apps/render-ios-playground/render-ios-playground/Application/UseCases/FetchScenarioUseCase.swift
import Foundation

/// Use case for fetching schema from remote source
class FetchScenarioUseCase {
    private let scenarioRepository: ScenarioRepository
    
    init(scenarioRepository: ScenarioRepository) {
        self.scenarioRepository = scenarioRepository
    }
    
    func execute(from url: URL) async throws -> Scenario? {
        do {
            return try await scenarioRepository.fetchScenario(from: url)
        } catch {
            if error is DomainError {
                throw error
            } else {
                throw ApplicationError.scenarioFetchFailed(error.localizedDescription)
            }
        }
    }
}

File: apps/render-android-playground/app/src/main/java/com/max_mrtnv/render_android_playground/application/usecases/FetchScenarioUseCase.kt
package com.max_mrtnv.render_android_playground.application.usecases

import com.max_mrtnv.render_android_playground.domain.entities.Scenario
import com.max_mrtnv.render_android_playground.domain.repositories.ScenarioRepository

/**
 * Use case for fetching scenario from remote source
 */
class FetchScenarioUseCase(
    private val scenarioRepository: ScenarioRepository
) {
    suspend fun execute(url: String): Scenario? {
        return scenarioRepository.fetchScenario(url)
    }
}
